### DISCUSSION TITLE
Is the test for parallel std::is_partitioned compatible with EXHAUSTIVE?

### DISCUSSION THREAD

**Discussion by mikekazakov** [microsoft/STL#3172] [General] [answered]
Hi STL devs!

There's a chance that the test at `STL/tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp` doesn't like the EXHAUSTIVE testing. When this macro is turned on, the iteration at `:105` with `testSize=3`  ends up with a container `{1, 0, 0}` which is partitioned, but the assert expects it to be not partitioned:

<img width="707" alt="Screenshot 2022-10-23 at 15 00 42" src="https://user-images.githubusercontent.com/7439287/197396514-a8f63c28-cd3c-4220-afc9-90ead7c4786c.png">

It's reproducible:
```
git clone https://github.com/microsoft/STL.git
sed -i -e 's/is_partitioned(par,/is_partitioned(/g' STL/tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp
sed -i -e 's/using namespace std::execution;//g' STL/tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp
sed -i -e 's/#include <execution>;//g' STL/tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp
clang++ -std=c++17 STL/tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp -ISTL/tests/std/include -o test && ./test && echo OK
clang++ -std=c++17 STL/tests/std/tests/P0024R2_parallel_algorithms_is_partitioned/test.cpp -ISTL/tests/std/include -o test -DEXHAUSTIVE && ./test && echo OK
> OK
> Assertion failed: (!is_partitioned( c.begin(), c.end(), read_char_as_bool)), function test_case_is_partitioned_parallel, file test.cpp, line 105.
```
Pardon the sed hackery - found this under macOS where the test fails even with the traditional serial implementation from libc++.

**Comment by CaseyCarter** [comment, marked as answer]
We don't normally run this test with `EXHAUSTIVE` defined, but it would be nice if doing so worked correctly. This case, for example, needs  `testSize >= 4` but isn't guarded appropriately. (The next case is fine with `testSize == 3`). There are also some redundant `std::`-qualifications in here that need cleaning up.

I'll submit a tiny PR with a fix / cleanup if the `EXHAUSTIVE` run I'm validating locally ever completes.

EDIT: I suppose the proper answer here is "No, this test is not compatible with `EXHAUSTIVE`, but there's no particular reason it shouldn't be."

**Reply by CaseyCarter** [reply]
Submitted https://github.com/microsoft/STL/pull/3174. The `EXHAUSTIVE` test run took only 15728.15s to complete on my 3950X - I wonder why we don't run this all the time? ;)

**Reply by mikekazakov** [reply]
Thank you a lot! Indeed Azure won't be very happy with running this test occasionally :)
And overall thanks for the great test suite, regarding both coverage and portability - kudos!
